Add MongoDB support for analytics records

jmalvarezf opened the pull request:
Hi,

As part of the integration of WSO2 in our company, El Corte Ingles, we have created a module to be able to read and write analytics events to a MongoDB database. Please take a look at the code, and share your comments about that. I have been able to go thourgh all the tests defined by using a mongodb docker instance.

I've sent an email with the design principles, but I can share them with you though this platform too. Any comments are, of course, very welcome.

Jose Maria.

chamilad commented:
@jmalvarezf Thanks for the PR! I couldn't find the email you mentioned about. It would be better to explain the user story in the PR itself too, to understand the design principles better.

jmalvarezf commented:
Sure, sorry for that, I've done this way in other PR, but I forgot to include the details in this one.
Basically, the principles are:

- All that is a table in the RDBMS world is going to be a collection in Mongo.
- There will be no support for partitioning inside the AnalyticsRecordStore, but left to sharding support in Mongo. We need to analyze the best key or compound key in the data to have better performance when writing, but this should be done apart from the code.
- We use the tenantId and the tableName to create the collection name, as you do in Cassandra implementation.
- The event will be saved with the default id created by mongo, and a field called arsId for the record ID field. The map with properties will be saved the same way it's saved in cassandra, leaving values as binaries.

The rest is pretty straightforward. We have tested it also in a real installation (two docker instances, one for the API-GW, one for the APIM-Analytics).

Best regards,

jmalvarezf commented:
There is something that you should also know. 
Right now, the development done does not support sharding. To support sharding, the user which we use to connect would have to have enablesharding privileges, and a modification should be done in the code to shard the collection every time we create one. It is not difficult to implement, given the appropiate rights, but it is not there. If you think this would be interesting, I could add that to the PR.

jmalvarezf commented:
This last commit tries to address a potential problem with a configuration that involves using writeConcern = 0 and readPreference = SECONDARY. It may happen that you create a collection (a table in the RDBMS world) and then you try to get information from there, but the collection may not exist if you are reading from a secondary that has not replicated the change.

Please, share your feedback with us about the PR.

Thanks.

gimantha commented:
@jmalvarezf I am creating the feature out of the component you created., since I couldn't find the feature in your pull request. Then we can actually install the mongo recordstores without copying the jars to /lib folder.. All we need to do is, including the feature in the p2-gen and build the product.

jmalvarezf commented:
@gimantha you are completely right sorry for that. I only built the package I needed and not the whole project. Thanks for your support!

lafernando commented on components/analytics-data-connectors/org.wso2.carbon.analytics.datasource.mongo/src/main/java/org/wso2/carbon/analytics/datasource/mongo/MongoAnalyticsRecordStore.java:
```diff
@@ -0,0 +1,250 @@
+        try {
+            Map<String, Document> documentsToInsert = new HashMap<String, Document>();
+            for (Record record : batch) {
+                if (db.getCollection(collection).count(eq("arsId", record.getId())) >= 1) {
```
Why don't we use "_id" as the record id? .. it is already the primary key of a document, and so it's automatically indexed also. And, here a count operation before every record insertion seems a bit expensive, why don't we try to do the batch insert straightaway, and if it fails, do an update/replaceOne for all the records of that batch.

jmalvarezf commented:
Hello @lafernando . I've done the following changes:

- I've added the license headers in all the files from this PR. I will add the header in the other PR involved later on.
- I've removed the author tag.
- I've changed the arsId field for the _id internal of Mongo. This was my first approach, but I had problems passing the tests with the "_id" (duplicated records and stuff). It seems that whatever problem I was experiencing is solved now on the code, so it is already refactored.
- I've changed the code to try to replace one document, and if the result was that no document was modified, then do an insert.

Thanks all,

jmalvarezf commented on components/analytics-data-connectors/org.wso2.carbon.analytics.datasource.mongo/src/main/java/org/wso2/carbon/analytics/datasource/mongo/AnalyticsRecord.java:
```diff
@@ -0,0 +1,121 @@
+package org.wso2.carbon.analytics.datasource.mongo;
```
Done

lafernando commented on components/analytics-data-connectors/org.wso2.carbon.analytics.datasource.mongo/src/main/java/org/wso2/carbon/analytics/datasource/mongo/MongoAnalyticsRecordStore.java:
```diff
@@ -0,0 +1,265 @@
+        try {
+            Map<String, Document> documentsToInsert = new HashMap<String, Document>();
+            for (Record record : batch) {
+                UpdateResult result = db.getCollection(collection).replaceOne((eq("_id", record.getId())),
```
Here, still problem is, for each record in the batch, you're doing a database operation to update the one if it's already there. This would be slow when you need to write the record batches efficiently. So what you need to do is, first, try to insert the whole batch, with a "insertMany", only if it fails, then fallback to a record by record "replaceOne". You can actually probably check the performance difference after doing this change.

jmalvarezf commented:
Hi @lafernando,

I've tried your approach, and it almost doubles the performance trying to insert directly in the tests. I think this approach makes sense if you do mostly inserts and you rarely do updates (I suppose that's the way the product works). I've modified it, and I only try replacing and then inserting if I got an bulk write exception. 
You can now check those modifications.

Best regards,

gimantha commented on components/analytics-data-connectors/org.wso2.carbon.analytics.datasource.mongo/pom.xml:
```diff
@@ -0,0 +1,105 @@
+	<dependency>
+            <groupId>org.mongodb</groupId>
+            <artifactId>mongo-java-driver</artifactId>
+            <version>3.4.1</version>
```
Lets move this to root pom.xml <dependancyManagement> and get a reference here.

gimantha commented on components/analytics-data-connectors/org.wso2.carbon.analytics.datasource.mongo/pom.xml:
```diff
@@ -0,0 +1,105 @@
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon.data</groupId>
```
Move this to root pom and get a reference here

jmalvarezf commented:
Hello @gimantha ,

I've moved the dependencies into the master pom of carbon-analytics. The carbon.data version property didn't match with the version that I need for the reader, so I created another property called "carbon.data.snapshot.version", with the appropiate version. If you want to change the name or whatever, let me know.

Thank you,

gimantha commented:
@jmalvarezf No problem,
I will update to the latest and merge.
Thank you

marcses commented:
Hi,
I tried to configure MongoDb as event store, but when I add the datasource to analytics-datasources.xml
 `<datasource>
           <name>WSO2_ANALYTICS_DS_MONGO</name>
           <description>The datasource used for analytics record store</description>
           <definition type="MONGO">
               <configuration>
                   <host>xxx.xxx.xxx.xxx</host>
                   <port>27017</port>
               </configuration>
           </definition>
		</datasource>`

and in analytics-config.xml :
`<analytics-record-store name="EVENT_STORE">
		<implementation>org.wso2.carbon.analytics.datasource.mongo.MongoAnalyticsRecordStore</implementation>
		<properties>
		   <property name="datasource">WSO2_ANALYTICS_DS_MONGO</property>
		   <property name="database">WSO2_DAS_EVENT</property>
		</properties>
	</analytics-record-store>`

I get the follow error:
`TID: [-1234] [] [2018-03-05 13:47:10,701]  WARN {org.wso2.carbon.metrics.impl.MetricServiceImpl} -  Failed to build the reporter {org.wso2.carbon.metrics.impl.MetricServiceImpl}
org.wso2.carbon.metrics.impl.util.ReporterBuildException: Error when looking up the Data Source: 'jdbc/WSO2MetricsDB'.
	at org.wso2.carbon.metrics.impl.util.JDBCReporterBuilder.build(JDBCReporterBuilder.java:174)
	at org.wso2.carbon.metrics.impl.util.JDBCReporterBuilder.build(JDBCReporterBuilder.java:33)
	at org.wso2.carbon.metrics.impl.MetricServiceImpl.<init>(MetricServiceImpl.java:201)
	at org.wso2.carbon.metrics.impl.MetricServiceImpl.<init>(MetricServiceImpl.java:65)
...
	at org.eclipse.osgi.framework.internal.core.StartLevelManager.doSetStartLevel(StartLevelManager.java:243)
	at org.eclipse.osgi.framework.internal.core.StartLevelManager.dispatchEvent(StartLevelManager.java:438)
	at org.eclipse.osgi.framework.internal.core.StartLevelManager.dispatchEvent(StartLevelManager.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:340)
Caused by: javax.naming.NameNotFoundException: Name [WSO2MetricsDB] is not bound in this Context. Unable to find [WSO2MetricsDB].
	at org.apache.naming.NamingContext.lookup(NamingContext.java:825)
	at org.apache.naming.NamingContext.lookup(NamingContext.java:159)
	at org.apache.naming.NamingContext.lookup(NamingContext.java:836)
	at org.apache.naming.NamingContext.lookup(NamingContext.java:173)
	at org.wso2.carbon.context.internal.CarbonContextDataHolder$CarbonInitialJNDIContext.lookup(CarbonContextDataHolder.java:709)
	at javax.naming.InitialContext.lookup(InitialContext.java:417)
	at org.wso2.carbon.metrics.impl.util.JDBCReporterBuilder.build(JDBCReporterBuilder.java:172)
	... 86 more`

Thank for any help
